src/Workspaces/Remote/ServiceHub/Shared/ServiceHubServiceBase.cs

```diff
@@ -166,86 +163,127 @@ private static Task<Solution> GetSolutionAsync(RoslynServices roslynService, Pin
             return solutionController.GetSolutionAsync(solutionInfo.SolutionChecksum, solutionInfo.FromPrimaryBranch, cancellationToken);
         }
 
+        protected async Task<T> RunServiceAsync<T>(Func<Task<T>> callAsync, CancellationToken cancellationToken)
```

Comment:
❓ Does StreamJsonRpc have events we can listen to before and after remoting calls? /cc @AArnott

Reply:
StreamJsonRpc has a `TraceSource` which we raise trace events on in many places, including just before dispatching an incoming request. You can search for `tracesource` [here](https://github.com/Microsoft/vs-streamjsonrpc/blob/master/src/StreamJsonRpc/JsonRpc.cs) to see a sampling, but there are more in other source files.

But if you're looking for a more formal eventing system where you can depend on specific timings, metadata on the event etc. surrounding a remoting call, I don't think we have that. Are you thinking of events around _incoming_ calls or outbound calls? Maybe file an issue on vs-streamjsonrpc with a feature request?

Does that answer your question?

Reply:
it would be nice if I can give some kind of lambda for incomming call and proxy call that will let json rpc to wrap every call under it.

Reply:
Incoming call I get: we basically pass you a delegate to call so you can wrap it however you like. Why is that? As far as handling exceptions, JsonRpc already has proper support for dealing with exceptions (at least, letting you decide which ones are "fatal" and should crash the process). Do you want to leverage that? As for logging, we could provide you an event if you want. Between those two features, do you still need to be able to actually invoke the method yourself?

You mentioned "and proxy call". Is you referring to the client proxy that StreamJsonRpc can generate for you? Something on the client side? Or is your request just about the server-side?

Reply:
yep. I am referring to Rpc.Attach<Proxy>(...) stuff on client side.

logging is one, but also, I have something I call every time I get call from host. I use that to detect idle. if service hub itself can provide like idle event, than I guess I can use that? like let me know if my servicehub process didn't get new request for 3 mins. something like that.

I use idle to clean caches and etc.

Reply:
tagging @AArnott

Reply:
It sounds like your idle check would be solved with an event. Our `JsonRpc.TraceSource` can tell you every time a call comes in. Just add a listener for it. There is [public API](https://github.com/Microsoft/vs-streamjsonrpc/blob/c00b81752ea45c35e1d8fde1ad1654575f2eb18c/src/StreamJsonRpc/JsonRpc.cs#L258-L261) that lets you detect specific trace events by their ID/type.